TST dtype dependent rtol, atol in assert_allclose_dense_sparse

oleksandr-pavlyk opened the pull request:
Check for closeness of new_result and the previously computed result
in `check_fit_idempotent` should depend on the dtype of the result, and 
cannot be less than that precision's epsilon.

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs

Closes #13977 

#### What does this implement/fix? Explain your changes.

This change sets `atol` keyword in the call to `assert_allclose_dense_sparse` in 
`check_fit_idempotent` estimator checker.

#### Any other comments?


@amueller @jeremiedbb

oleksandr-pavlyk committed:
MAINT: Per #13977 set non-default absolute tolerance

Check for closeness of new_result and the previously compute result
should depend on the dtype of the result, and can not be less than
that precision's epsilon.

oleksandr-pavlyk commented:
is this build hung?

oleksandr-pavlyk commented:
@ogrisel @jeremiedbb CI is green. Could you review, please.

jeremiedbb commented:
Shouldn't we adjust the `rtol` instead ? unless comparing to 0, atol is usually not meaningful and kind of unrelated to the dtype.

oleksandr-pavlyk commented:
Yes, if that case how about setting `atol = rtol = np.finfo(dtype).epsilon` ?

jeremiedbb commented:
Yes that's fine

oleksandr-pavlyk committed:
set both atol and rtol to dtype's epsilon

oleksandr-pavlyk commented:
@jeremiedbb Done.

jeremiedbb commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2448,4 +2448,12 @@ def check_fit_idempotent(name, estimator_orig):
+            if np.issubdtype(new_result.dtype, np.floating):
+                tol = np.finfo(new_result.dtype).eps
+            else:
+                tol = np.finfo(np.float64).eps
```
maybe add a comment `# integer dtype` because it's not clear at first sight.

I wonder if it would not be even clearer if you reformulate this way:
```python
if new_result.dtype == np.float32:
    tol = np.finfo(np.float32).eps
else:
    tol = np.finfo(np.float64).eps
```
wdyt ?

oleksandr-pavlyk replied:
I think using np.floating is much cleaner NumPy, it accounts for np.half, np.longdouble etc., even though these types may not be used in Scikit-Learn.

jeremiedbb replied:
fair enough :)

jeremiedbb reviewed (approved):
just a minor comment. Other than that lgtm.

oleksandr-pavlyk commented:
pinging reviewers, please.

rth reviewed (dismissed):
LGTM. Thanks @oleksandr-pavlyk !

rth commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2448,4 +2448,12 @@ def check_fit_idempotent(name, estimator_orig):
             new_result = getattr(estimator, method)(X_test)
-            assert_allclose_dense_sparse(result[method], new_result)
+            if np.issubdtype(new_result.dtype, np.floating):
+                tol = np.finfo(new_result.dtype).eps
```
No actually for float64, this would `2.22e-16` do we actually need such precision? I think the previous default of 1e-7 was enough? For other dtypes, I agree.

Maybe then,
```
rtol = max(tol, 1e-7)
atol = tol
```

also maybe take `2*np.finfo(dtype).eps` just to be safe?

oleksandr-pavlyk replied:
Thanks for the good feedback. I agree, and have pushed both proposed changes.

oleksandr-pavlyk committed:
Implemented PR feedback

Parameter tol is set to be 2*np.finfo(dtype).eps, rather than
np.finfo(dtype).eps.

Setting of rtol and atol parameters of `assert_allclose_dense_sparse`
is modified to not go below the default values.

rth commented:
Thanks!
